src/main/software/amazon/event/ruler/GenericMachine.java

```diff
@@ -454,23 +500,29 @@ private void addStep(final NameState state,
         // for each pattern, we'll provisionally add it to the BMC, which may already have it.  Pass the states
         //  list in in case the BMC doesn't already have a next-step for this pattern and needs to make a new one
         //
+        NameState lastNextState = null;
+        Set<NameState> nameStates = new HashSet<>();
         for (Patterns pattern : patterns.get(key)) {
-            final NameState nextState;
-
             if (isNamePattern(pattern)) {
                 assert nameMatcher != null;
-                nextState = nameMatcher.addPattern(pattern, NameState::new);
+                final NameState nameStateForSupplier = lastNextState == null ? new NameState() : lastNextState;
+                lastNextState = nameMatcher.addPattern(pattern, () -> nameStateForSupplier);
             } else {
                 assert byteMachine != null;
-                nextState = byteMachine.addPattern(pattern);
+                lastNextState = byteMachine.addPattern(pattern, lastNextState);
```

Comment:
Yes, good observation that the order the rules are added matters. I didn't solve this part. To be clear - are you pointing this out as an FYI or do you have a suggestion? I'm thinking a bit and could see solving this with some sort of a NameStateRegistry per GenericMachine. This registry would vend a NameState given a list of keys (up to a certain index). Feels like it could be a big win. Let me think about this a bit more too.

Reply:
this is FYI as I don't have simple to resolve it unless introduce the name state merge logic which seems a chain effect and looks very complex. 
I think with this change, further change of memory optimization become low priority.